interfaces/cython/cantera/ck2yaml.py

```diff
@@ -0,0 +1,2111 @@
+                            self.warn('"THERMO" section implicitly ended by start of '
+                                      'next section on line {0}.'.format(self.line_number))
+                            advance = False
+                            tokens.pop()
+                            break
+
+                        if comment:
+                            current.append('!'.join((line, comment)))
+                        else:
+                            current.append(line)
+                        if len(line) >= 80 and line[79] in ['1', '2', '3', '4']:
+                            thermo.append(line)
+                            if line[79] == '4':
+                                try:
+                                    label, thermo, comp = self.read_NASA7_entry(thermo, TintDefault, comments)
+                                except Exception as e:
+                                    error_line_number = self.line_number - len(current) + 1
+                                    error_entry = ''.join(current).rstrip()
+                                    logging.info(
+                                        'Error while reading thermo entry starting on line {0}:\n'
+                                        '"""\n{1}\n"""'.format(error_line_number, error_entry)
+                                    )
+                                    raise
+
+                                if label not in self.species_dict:
+                                    if skip_undeclared_species:
+                                        logging.info('Skipping unexpected species "{0}" while reading thermodynamics entry.'.format(label))
+                                        thermo = []
+                                        line, comment = readline()
+                                        current = []
+                                        comments = [comment]
+                                        continue
+                                    else:
+                                        # Add a new species entry
+                                        species = Species(label=label)
+                                        self.species_dict[label] = species
+                                        self.species_list.append(species)
+                                else:
+                                    species = self.species_dict[label]
+
+                                # use the first set of thermo data found
+                                if species.thermo is not None:
+                                    self.warn('Found additional thermo entry for species {0}. '
+                                              'If --permissive was given, the first entry is used.'.format(label))
+                                else:
+                                    species.thermo = thermo
+                                    species.composition = comp
+
+                                thermo = []
+         
```

Comment:
Because borrowing the existing implementation of the Chemkin-reading code from `ck2cti` was a more efficient use of time than rewriting it.

Reply:
Ha! Fair 😄 I think it would be worth leaving a comment or filing an issue (this would actually make a good first issue for someone who's familiar with Python) to refactor this into its own function.

Reply:
I think there are many more valuable contributions that any potential contributor could make than refactoring this code.

Reply:
Perhaps, but this one doesn't necessarily require specific knowledge of esoteric formats or detailed chemistry. It's probably not worth our (meaning yours, mine, or Steven's) time, but new people have to get started *somewhere*. For instance, this could be a good thing for a new grad student to do to get them familiar with 1) Cantera's code base, 2) setting up a Cantera dev environment and running the tests, 3) git, 4) GitHub & PRs and expect them to finish in a day or two.

Reply:
Refactoring this loop is *entirely* about understanding an esoteric format, and one without complete or accurate documentation at that. I think it would take far longer than a couple of days to refactor this code without inadvertently breaking some of the bizarre special cases that are allowed by the format. And while the test suite for this is substantial, I guarantee that it is not complete.

In any case, I have added a "TODO" note which at least makes it clear that there is room for improvement in this code. I'd rather not start creating Issues for "this chunk of code could be better".